### DISCUSSION TITLE
Cypress Cloud only: shows the "404" page for NextJS dynamic routes with 'force-static'. How to debug?

### DISCUSSION THREAD

**Discussion by IonelLupu** [cypress-io/cypress#28403] [Questions and Help]
I have a NextJS app and a couple of e2e tests. All the tests that hit the pages that are under a dynamic route name and contain `export const dynamic = 'force-static'` return the "not found" page. The same exact page (with the same exact source code) that is not under a dynamic route work correctly in cypress cloud.

One note: all the tests pass when running on my local machine.

For example, here is the file tree of my project:

```
app
└── hotels
    ├── [id]
    │   └── [slug]
    │       └── page.tsx  // tests fail for this page on cypress cloud only
    │   ├── another
    │       └── page.tsx  // tests fail for this page on cypress cloud only
    ├── another
    │   └── page.tsx  // tests pass here
    └── test
        ├── [test]
        │   ├── another
        │   │   └── page.tsx  // tests fail for this page on cypress cloud only
        │   └── another
        │       └── page.tsx   // tests pass here
```

The test suite is very simple:
```ts
context(`some test context`, { tags: ['@mobile', '@desktop', '@developmentApi'] }, () => {
    it('home page', () => {
        cy.visit('/')
        cy.contains('Hello world')
    })
    it('hotel slug another', () => {
        cy.visit('/hotels/18923395/another')
        cy.contains('Hello world')
    })
    it('hotel another', () => {
        cy.visit('/hotels/another')
        cy.contains('Hello world')
    })
    it('hotel test another', () => {
        cy.visit('/test/another')
        cy.contains('Hello world')
    })
    it('hotel test slug another', () => {
        cy.visit('/test/random/another')
        cy.contains('Hello world')
    })
    it('hotel original', () => {
        cy.visit('/hotels/18923395/Hotel-Parador')
        cy.contains('Hello world')
    })
})

export {}


```

With all of these, those 3 tests fail to pass on Cypress Cloud only saying the page was not found:
![image](https://github.com/cypress-io/cypress/assets/4083652/f786d0d3-483c-43fb-8214-442f2c3a75ea)

And every page.tsx content is this:
```ts
import React from 'react'

export const dynamic = 'force-static'
export const revalidate = 1800

export default async function HotelPage() {
    return <>Hello world</>
}

```

How can I debug this in cypress cloud? I don't have access to the NextJS server to see the logs

--- 

### **Edit**
After days of debugging, it seems building the project with NODE_ENV=test is the real issues: https://github.com/vercel/next.js/issues/58935

**Comment by portyu9** [comment]
Hi @IonelLupu 

Your edit identified the key point: this is not a Cypress Cloud routing problem.

The same behavior was reproduced in Next.js issue `#58935` with:

```bash
NODE_ENV=test next build
```

and a dynamic App Router route using:

```js
export const dynamic = 'force-static'
```

The fix is to build the deployable Next.js application using its normal production environment:

```bash
next build
```

or explicitly:

```bash
NODE_ENV=production next build
```

Then run Cypress against that production build.

`NODE_ENV=test` is appropriate for the **test process**, but it should not be used to change a production `next build` into a test-mode build.

If you need Cypress-specific application configuration, use a separate environment variable:

```bash
APP_ENV=test next build
```

and keep:

```text
NODE_ENV=production
```

for the production build.

Also, Cypress Cloud does not host your Next.js application. Cypress runs in your CI environment and records the results to Cloud, so the Next.js server/build logs need to be captured by your CI job rather than retrieved from Cypress Cloud itself.

A useful CI sequence is:

```bash
npm run build
npm run start &
npx wait-on http://localhost:3000
npx cypress run --record
```

If a page returns a 404 there, inspect the build/server output from the CI job.

Please mark this answer as accepted if it helped, thank you.
